modules/users/tests/server/user.server.routes.tests.js

```diff
@@ -1086,6 +1086,118 @@ describe('User CRUD tests', function () {
     });
   });
 
+  describe('Username Update', function () {
+    it('should update it', function (done) {
+      var user2 = _user;
+      user2.username = _user.username + '01';
+      delete user2.email;
+      agent.post('/api/auth/signin')
+        .send(credentials)
+        .expect(200)
+        .end(function (err) {
+          if (err) {
+            return done(err);
+          }
+          agent.put('/api/users')
+            .send(user2)
+            .expect(200)
+            .end(function (err, res) {
+              if (err) {
+                return done(err);
+              }
+              res.body.username.should.equal(user2.username);
+              return done();
+            });
+        });
+    });
+
+    it('should fail to update if the three months have not passed by',
+      function (done) {
+        var user2 = _user;
+        delete user2.email;
+        agent.post('/api/auth/signin')
+          .send(credentials)
+          .expect(200)
+          .end(function (err) {
+            if (err) {
+              return done(err);
+            }
+            // First change
+            user2.username = _user.username + '01';
+            agent.put('/api/users')
+              .send(user2)
+              .expect(200)
+              .end(function (err, res) {
+                if (err) {
+                  return done(err);
+                }
+                res.body.username.should.equal(user2.username);
+                user2.username = _user.username + '02';
+                agent.put('/api/users')
+                  .send(user2)
+                  .end(function (err, res) {
+                    if (err) {
+                      return done(err);
+                    }
+                    res.body.message.should.equal('You have used your chance to change your username already');
+                    return done();
+                  });
+              });
+          });
+      });
+
+    it('should update if the three months rule have passed by',
+      function (done) {
+        var user2 = _user;
+        delete user2.email;
+        agent.post('/api/auth/signin')
+          .send(credentials)
+          .expect(200)
+          .end(function (err) {
+            if (err) {
+              return done(err);
+            }
+            // First change
+            user2.username = _user.username + '01';
+            agent.put('/api/users')
+              .send(user2)
+              .expect(200)
+              .end(function (err, res) {
+                if (err) {
+                  return done(err);
+                }
+                res.body.username.should.equal(user2.username);
+                User.findById(user._id, function (err, user) {
+                  if (err) {
+                    return done(err);
+                  }
+                  var threeMonthsAgo = new Date(user.usernameUpdated);
+                  threeMonthsAgo.setMonth(
```

Comment:
It could be nice to use the `sinon` library for testing the 3-month period. It would make the tests more independent from the implementation.

It could look something like this (untested):
```js
var sinon = require('sinon');

describe(function () {
  var sandbox;

  beforeEach(function () { // this block gets executed before every test within the same describe block
    sandbox = sinon.sandbox.create(); // sandbox makes it easy to restore the environment after the tests
    sandbox.useFakeTimers(1500000000000, 'Date'); // sets current date to a specific timestamp (milliseconds)
    // the 1500...00 timestamp is my personal favorite. Any reasonable timestamp will do. :)
  });

  afterEach(function () { // executed after every test within this describe block
    sandbox.restore(); // restores the test environment
  });

  it('test description', function () {
    ...
    // send the first request
    // test success
    ...
    sandbox.clock.tick(<3 months in milliseconds>); // wait
    ...
    // send the second request
    // test success
  })
})
```

A full example of using sinon with fake timers can be found in https://github.com/Trustroots/trustroots/blob/21c7f3ee7ffd5705bba39404c738429dc3c13144/modules/users/tests/server/user-lastseen.server.tests.js
(full disclosure: i wrote those tests)

Sinon is a very useful tool for other test-related stuff, too!

Reply:
You told me about the lib before. I just though that testing a real thing and I mean by that having the date set at the database would be more interesting...

Reply:
Note that our version of Sinon is a bit outdated (PRs welcome to update it!)